clippy_lints/src/casts/mod.rs

```diff
@@ -754,6 +755,31 @@ declare_clippy_lint! {
     "detects `as *mut _` and `as *const _` conversion"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for casts to a pointer of type `T`, from `mem::align_of::<U>()` such that `T` = `U`,
+    /// or from a constant address such that it matches `T`'s alignment.
+    ///
+    /// ### Why is this bad?
+    /// This creates a pointer with no provenance which might cause incorrect compiler optimizations.
+    /// It is better expressed as {`std`, `core`}`::ptr::`{`dangling`, `dangling_mut`}.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a = 4 as *const u32;
+    /// let b = std::mem::align_of::<u32>() as *const u32;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a = std::ptr::dangling::<u32>();
+    /// let b = std::ptr::dangling::<u32>();
+    /// ```
+    #[clippy::version = "1.86.0"]
+    pub DANGLING_PTR,
```

Comment:
The lint name would be better as `manual_dangling_ptrs`. Just naming it `dangling_ptr` makes it seem like this lints against `ptr::dangling`.